wire/merkle_tree.go

```diff
@@ -0,0 +1,310 @@
+package wire
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/tokenized/pkg/bitcoin"
+)
+
+// MerkleTree is an efficient structure for calculating a merkle root hash.
+type MerkleTree struct {
+	layers []*merkleNodeLayer // First layer, index zero, is the lowest level of the tree.
+	prune  bool
+	count  int
+
+	merkleProofs []*MerkleProof
+}
+
+func NewMerkleTree(prune bool) *MerkleTree {
+	return &MerkleTree{
+		prune: prune,
+	}
+}
+
+// AddMerkleProof adds a merkle proof to be calculated with the tree.
+func (t *MerkleTree) AddMerkleProof(txid bitcoin.Hash32) {
+	t.merkleProofs = append(t.merkleProofs, NewMerkleProof(txid))
+}
+
+func (t MerkleTree) Print() {
+	indent := ""
+	for i, layer := range t.layers {
+		fmt.Printf("%sLayer %d\n", indent, i+1)
+		indent += "  "
+		hashIndex := layer.count - len(layer.hashes) + 1
+		for _, hash := range layer.hashes {
+			fmt.Printf("%s%d : %s\n", indent, hashIndex, hash.String())
+			hashIndex++
+		}
+	}
+}
+
+// AddHash adds a new hash to the merkle tree.
+func (t *MerkleTree) AddHash(hash bitcoin.Hash32) {
```

Comment:
SonarCloud gives this 19 out of 15. Some fairly easy ways to reduce the complexity.

Reply:
Moved this functions hash calculation block out too.